ttl/sql.go

```diff
@@ -0,0 +1,468 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ttl
+
+import (
+	"encoding/hex"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/parser/format"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/parser/mysql"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/sqlexec"
+	"github.com/pkg/errors"
+)
+
+func writeHex(in io.Writer, d types.Datum) error {
+	_, err := fmt.Fprintf(in, "x'%s'", hex.EncodeToString(d.GetBytes()))
+	return err
+}
+
+func writeDatum(in io.Writer, d types.Datum, ft *types.FieldType) error {
+	switch d.Kind() {
+	case types.KindString, types.KindBytes, types.KindBinaryLiteral:
+		if mysql.HasBinaryFlag(ft.GetFlag()) {
+			return writeHex(in, d)
+		}
+		_, err := fmt.Fprintf(in, "'%s'", sqlexec.EscapeString(d.GetString()))
+		return err
+	}
+	ctx := format.NewRestoreCtx(format.DefaultRestoreFlags, in)
+	expr := ast.NewValueExpr(d.GetValue(), ft.GetCharset(), ft.GetCollate())
+	return expr.Restore(ctx)
+}
+
+// FormatSQLDatum formats the datum to a value string in sql
+func FormatSQLDatum(d types.Datum, ft *types.FieldType) (string, error) {
+	var sb strings.Builder
+	if err := writeDatum(&sb, d, ft); err != nil {
+		return "", err
+	}
+	return sb.String(), nil
+}
+
+type sqlBuilderState int
+
+const (
+	writeBegin sqlBuilderState = iota
+	writeSelOrDel
+	writeWhere
+	writeOrderBy
+	writeLimit
+	writeDone
+)
+
+// SQLBuilder is used to build SQLs for TTL
+type SQLBuilder struct {
+	tbl   *PhysicalTable
+	sb    strings.Builder
```

Comment:
Maybe you should consider https://github.com/pingcap/tidb/blob/d100e933c8982ed13135abbca1224b472b167b53/util/sqlexec/utils.go#L247
if there is a risk of injection.

Reply:
I'm using `sqlexec.EscapeString` to escape the value. Seems they are equivalent in implement

Reply:
Sure , but table name, or every identifier may have a problem. BTW, `FormatSQL` is the builder version of `EscapeSQL`. Check again if you really need escaping everywhere or just datum.